src/diffusers/models/vq_diffusion_attention.py

```diff
@@ -0,0 +1,246 @@
+                Conditional embeddings for cross attention layer.
+            t (:obj: `torch.long`):
+                Denoising timestep
+
+        Returns:
+            `torch.FloatTensor` of shape `(batch size, num embed - 1, num latent pixels)`:
+                Probability distributions for the unnoised latent pixels. Note that it does not output a prediction for
+                the masked class.
+        """
+        embedded_latent_images = self.latent_image_embedding(latent_images)
+        hidden_states = embedded_latent_images
+
+        for block in self.transformer_blocks:
+            hidden_states = block(hidden_states, cond_emb, t)
+
+        logits = self.out(self.norm_out(hidden_states))
+        # (batch, self.num_embed - 1, self.num_latent_pixels)
+        logits = logits.permute(0, 2, 1)
+
+        log_p_x_0 = F.log_softmax(logits.double(), dim=1).float()
+
+        return log_p_x_0
+
+
+class VQDiffusionImageEmbeddings(nn.Module):
+    """
+    Converts latent image classes into vector embeddings for the transformer. Sums the vector embeddings with
+    positional embeddings for the height and width of the latent space.
+
+    Note that the vector embeddings for the transformer are different than the vector embeddings from the VQVAE.
+
+    Args:
+        num_embed (`int`):
+            Number of embeddings for the latent pixels embeddings.
+        height (`int`):
+            Height of the latent image i.e. the number of height embeddings.
+        width (`int`):
+            Width of the latent image i.e. the number of width embeddings.
+        embed_dim (`int`):
+            Dimension of the produced vector embeddings. Used for the latent pixel, height, and width embeddings.
+    """
+
+    def __init__(
+        self,
+        num_embed: int,
+        height: int,
+        width: int,
+        embed_dim: int,
+    ):
+        super().__init__()
+
+        self.height = height
+        self.width = width
+        self.num_embed = num_embed
+        self.embed_dim = embed_dim
+
+        self.emb = nn.Embedding(self.num_embed, embed_dim)
+        self.height_emb = nn.Embedding(self.height, embed_dim)
+        self.width_emb = nn.Embedding(self.width, embed_dim)
+
+    def forward(self, index):
+        emb = self.emb(index)
+
+        height_emb = self.height_emb(torch.arange(self.height, device=index.device).view(1, self.height))
+
+        # 1 x H x D -> 1 x H x 1 x D
+        height_emb = height_emb.unsqueeze(2)
+
+        width_emb = self.width_emb(torch.arange(self.width, device=index.device).view(1, self.width))
+
+        # 1 x W x D -> 1 x 1 x W x D
+        width_emb = width_emb.unsqueeze(1)
+
+        pos_emb = height_emb + width_emb
+
+        # 1 x H x W x D -> 1 x L xD
+        pos_emb = pos_emb.view(1, self.height * self.width, -1)
+
+        emb = emb + pos_emb[:, : emb.shape[1], :]
+
+        return emb
+
+
+class VQDiffusionTransformerBlock(nn.Module):
```

Comment:
This is quite a normal attention block - could we maybe re-use https://github.com/huggingface/diffusers/blob/888468dd90b4e353225d7d505fa28d2963e65678/src/diffusers/models/attention.py#L168 here?